libstorage/libstoragemgmt
# [PATCH V6 00/18] Pool RAID info query and hardware RAID volume creation.

* Putting two patch sets into one. As `pool_member_info()` patch set is
  just preparation for `volume_raid_create()` patch set. It's better to
  explain design choice with both.

* New method `pool_member_info()` to query pool member information:
    * RAID type
    * Member type: disk RAID pool or sub-pool.
    * Members: disks or parent pools.

* New method `volume_raid_create()` to create RAID group volume.

* New method `volume_raid_create_cap_get()` to query RAID group volume create
  supported RAID type and strip sizes.

* Design notes:
    * Why we need pool_member_info() when we already have volume_raid_info()?
        The `pool_member_info()` is focus on RAID disk or sub-pool layout.

        The `volume_raid_info()` is focus on providing information about host
        I/O performance and availability(raid).

    * Why don't create single sharable method for SAN/NAS/DAS about RAID
      creation?
        Please refer to design note of patch 'New feature: Create RAID volume
        on hardware RAID.'

    * Why no RAID creation method for SAN and NAS?
        As we tried before, due to complex RAID settings of SAN and NAS
        storage, I didn't find a simple design to support all of them.

    * Please refer to design note of patch 'New feature: Create RAID volume'
      also.

* Please comment in github page, thank you.
    https://github.com/libstorage/libstoragemgmt/pull/11

Changes in V3:

* Rename pool_raid_info to pool_member_info.
* Rebased patches for the rename.
* Fix C code warning about comparing signed and unsigned integer.
* Initialized pointers in lsm_plugin_ipc.cpp and tester.c.
* Removed unused variables in tester.c.
* volume_create_raid() tested on:
    * MegaRAID: RAID 0/1/5/10.  # No RAID 6 license.
    * HP SmartArray: RAID 0.    # No additional disk.
* Document pull request: https://github.com/libstorage/libstoragemgmt-doc/pull/3

Changes in V4:

* Rename method 'volume_create_raid' to 'volume_raid_create'.
* Fix memory leak in C API.
* Initialize pointers in C code.
* Make sure output pointer is set to sane value if error occurred in C code.
* Add bash auto completion for new lsmcli commands and aliases.
* Fix typo in comments of C and Python library.
* Use hash table to convert vendor RAID string to LSM type in hpsa and
  megaraid plugins.
* Tested MegaRAID for 4 disks RAID 10 with default strip size.
  Query commands is tested on MegaRAID and Smart Array.
* Document updated and C API user guide for these new methods also included:
    https://github.com/libstorage/libstoragemgmt-doc/pull/3

Changes in V5:

* Updated commit message to use current method name:
    * pool_member_info
    * volume_raid_create
    * volume_raid_create_cap_get

Changes in V6:

* Patch 1/18, 8/18:
    Fix memory leaks in C library and C unit test.

* Patch 18/18:
    New patch for fixing volume_raid_create C unit test.
